### PR TITLE
emit polymorphism model

### PR DESCRIPTION
**Pull request by chunyu3**
# Description
Fix issue: https://github.com/Azure/autorest.csharp/issues/2484

- emit polymorphism model

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

File: src/CADL.Extension/Emitter.Csharp/src/lib/model.ts [hunk 1]
```diff
@@ -426,6 +427,22 @@ export function getInputType(
...
+                    Name: discriminator.propertyName,
+                    SerializedName: discriminator.propertyName,
+                    Description: "",
+                    Type: {
+                        Name: "String",
+                        Kind: InputTypeKind.String,
+                        IsNullable: false
+                    } as InputPrimitiveType,
+                    IsRequired: true,
+                    IsReadOnly: false,
```
Comment (lirenhe) [inline src/CADL.Extension/Emitter.Csharp/src/lib/model.ts:441]: Why readonly is false?  Can we overwrite the value of a discriminator?

File: src/CADL.Extension/Emitter.Csharp/src/lib/model.ts [hunk 2]
```diff
@@ -426,6 +427,22 @@ export function getInputType(
...
+                    IsDiscriminator: true
+                };
+                properties.push(discriminatorProp);
```
Comment (lirenhe) [inline src/CADL.Extension/Emitter.Csharp/src/lib/model.ts:444]: If the discriminator is related to the convinience API, better to verify all derived models are also generated.

**Review by lirenhe — COMMENTED** [review]
This PR is only for the emitter work for discriminator.
Could you update the PR description to call out codegen part in this PR is just a place holder.

File: src/CADL.Extension/Emitter.Csharp/src/lib/model.ts (hunk 2 above)
Comment (chunyu3) [reply src/CADL.Extension/Emitter.Csharp/src/lib/model.ts:444]: I will collect usages for derived models in other PR.

File: test/AutoRest.Shared.Tests/AutoRest.Shared.Tests.csproj [hunk 3]
```diff
@@ -1,7 +1,7 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>net461;netcoreapp3.1;net60</TargetFrameworks>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
```
Comment (archerzz) [inline test/AutoRest.Shared.Tests/AutoRest.Shared.Tests.csproj:4]: Is this change expected?

File: src/CADL.Extension/Emitter.Csharp/src/lib/model.ts [hunk 4]
```diff
@@ -426,6 +427,22 @@ export function getInputType(
 
             models.set(name, model);
 
+            if(discriminator) {
+                const discriminatorProp = {
+                    Name: discriminator.propertyName,
+                    SerializedName: discriminator.propertyName,
+                    Description: "",
```
Comment (archerzz) [inline src/CADL.Extension/Emitter.Csharp/src/lib/model.ts:434]: maybe we should set the `Description` like `discriminator property for inheritance`?

File: src/CADL.Extension/Emitter.Csharp/src/lib/model.ts (hunk 1 above)
Comment (chunyu3) [reply src/CADL.Extension/Emitter.Csharp/src/lib/model.ts:441]: We need to provide value for discriminator parameter when we call the api

File: test/AutoRest.Shared.Tests/AutoRest.Shared.Tests.csproj (hunk 3 above)
Comment (chunyu3) [reply test/AutoRest.Shared.Tests/AutoRest.Shared.Tests.csproj:4]: No. I reverted it back.

**Review by lirenhe — APPROVED** [review]

File: src/CADL.Extension/Emitter.Csharp/src/lib/model.ts (hunk 4 above)
Comment (chunyu3) [reply src/CADL.Extension/Emitter.Csharp/src/lib/model.ts:434]: @archerzz  No. in cadl `discriminator` is not defined as an property of base model, it just a decorator. There is no doc for it. So we cannot set Description for it.

**Review by archerzz — APPROVED** [review]

**Pull request merged**